### PR TITLE
add snapshot support

### LINKED ISSUE #413: Snapshot management
Heketi should have a snapshot management API

### PR DESCRIPTION
**Pull request by MartinForReal**
### What does this PR achieve? Why do we need it?

implementing snapshot management functions. With this patch heketi can manage snapshot of glusterfs volumes

### Does this PR fix issues?

Fixes #413 

### Notes for the reviewer

I didn't write any code for python client library.

### CODE REVIEW

**Comment by phlogistonjohn** [comment]
Well, this is a pleasant surprise. :-)

I haven't gone too deeply into the the individual changes & before we do I'd like to request that you break the single commit into a short series of commits. For example, changes to test scripts should be their own independent commit. Individual changes. The golang code can be broken up a bit in that you can make changes where code to create an manage data types goes in one commit and then operations changes that relies on that code goes in another and then finally the rest api's are added in a 3rd commit (for example).

In addition, we're going to (eventually) need some unit and functional tests for the new features.

If you have questions about these requests, please just ask.

**Comment by phlogistonjohn** [comment]
I think that @nixpanic may be interested in looking at this when the time comes.

**Comment by MartinForReal** [comment]
@nixpanic Could you please help review the change set? Any comments is appreciated. Thanks!

**Comment by MartinForReal** [comment]
hi ~ can anybody help on this issue? Thanks in advance!

File: apps/glusterfs/operations.go [hunk 1]
```diff
@@ -659,6 +659,191 @@ func (vc *VolumeCloneOperation) Finalize() error {
...
+	OperationManager
+	noRetriesOperation
+	vol      *VolumeEntry
+	snapshot *SnapshotEntry
+}
+
+// NewVolumeSnapshotOperation returns a new VolumeSnapshotOperation populated
+// with the given volume entry and db connection and allocates a new
+// pending operation entry.
+
```
Comment (nixpanic) [inline apps/glusterfs/operations.go]: nit: unneeded empty line

File: apps/glusterfs/snapshot_entry.go [hunk 2]
```diff
@@ -0,0 +1,130 @@
+package glusterfs
+
+import (
+	"bytes"
+	"encoding/gob"
+	"github.com/boltdb/bolt"
```
Comment (nixpanic) [inline apps/glusterfs/snapshot_entry.go]: nit: we mostly split groups of imports with empty lines (go internal, external)

File: apps/glusterfs/snapshot_entry.go [hunk 3]
```diff
@@ -0,0 +1,130 @@
...
+func (s *SnapshotEntry) Delete(tx *bolt.Tx) error {
+	return EntryDelete(tx, s, s.Info.Id)
+}
+
+// Visible returns true if this snapshot is meant to be visible to
+// API calls.
+func (s *SnapshotEntry) Visible() bool {
+	return s.Pending.Id == ""
+}
+func (s *SnapshotEntry) NewInfoResponse(tx *bolt.Tx) (*api.SnapshotInfoResponse, error) {
```
Comment (nixpanic) [inline apps/glusterfs/snapshot_entry.go]: nit: missing empty line

File: apps/glusterfs/volume_snapshot.go [hunk 4]
```diff
@@ -0,0 +1,53 @@
+//
+// Copyright (c) 2015 The heketi Authors
```
Comment (nixpanic) [inline apps/glusterfs/volume_snapshot.go]: nit: it is 2018 already

File: apps/glusterfs/snapshot_entry.go [hunk 5]
```diff
@@ -0,0 +1,130 @@
+package glusterfs
```
Comment (nixpanic) [inline apps/glusterfs/snapshot_entry.go]: Needs copyright + license header.

File: client/api/go-client/snapshot.go [hunk 6]
```diff
@@ -0,0 +1,109 @@
+package client
```
Comment (nixpanic) [inline client/api/go-client/snapshot.go]: Needs copyright + license header

File: apps/glusterfs/volume_snapshot.go [hunk 7]
```diff
@@ -0,0 +1,53 @@
...
+// This file is licensed to you under your choice of the GNU Lesser
+// General Public License, version 3 or any later version (LGPLv3 or
+// later), or the GNU General Public License, version 2 (GPLv2), in all
+// cases as published by the Free Software Foundation.
+//
+package glusterfs
+
+import (
+	"errors"
+	"github.com/boltdb/bolt"
```
Comment (nixpanic) [inline apps/glusterfs/volume_snapshot.go]: nit: imports with empty lines encouraged

File: apps/glusterfs/volume_entry.go [hunk 8]
```diff
@@ -959,6 +963,15 @@ func (v *VolumeEntry) cloneVolumeRequest(db wdb.RODB, clonename string) (*execut
...
 
+	var sshhost string
+	sshhost, err := v.GetManageHost(db)
+	if err != nil {
+		return nil, "", err
+	}
+	return vcr, sshhost, nil
+}
+
+func (v *VolumeEntry) GetManageHost(db wdb.RODB) (string, error) {
```
Comment (nixpanic) [inline apps/glusterfs/volume_entry.go]: this function can probably be private? Just `getManageHost()` would do in that case.

File: client/api/go-client/snapshot.go [hunk 9]
```diff
@@ -0,0 +1,109 @@
...
+
+import (
+	"github.com/heketi/heketi/pkg/glusterfs/api"
```
Comment (nixpanic) [inline client/api/go-client/snapshot.go]: nit: re-order and group the imports

File: client/cli/go/cmds/snapshot.go [hunk 10]
```diff
@@ -0,0 +1,144 @@
+//
+// Copyright (c) 2018 The heketi Authors
```
Comment (nixpanic) [inline client/cli/go/cmds/snapshot.go]: Yay!

File: docs/api/api.md [hunk 11]
```diff
@@ -712,6 +717,115 @@ When a volume is deleted, Heketi will first stop, then destroy the volume.  Once
...
 ```
 
+### Create a Snapshot
+* **Method:** _POST_
+* **Endpoint**:`/volumes/{volume_uuid}/snapshot`
+* **Content-Type**: `application/json`
+* **Response HTTP Status Code**: 202, See [Asynchronous Operations](#asynchronous-operations)
+* **Temporary Resource Response HTTP Status Code**: 303, `Location` header will contain `/snapshots/{snapshot_uuid}`. See [Snapshot Info](#snapshot_info) for JSON response.
+* **JSON Request**:
+    * name: _string_, _optional_, Name of snapshot. If not provided, the name of the snapshot will be `snap_{id}`, for example `snap_728faa5522838746abce2980`
```
Comment (nixpanic) [inline docs/api/api.md]: I thought you used `snapshot_` as prefix?

File: tests/020-test-shellscripts.sh [hunk 12]
```diff
@@ -19,7 +19,7 @@ test_shellcheck() {
 	fi
 }
 
-SHELLCHECK="$(which shellcheck 2>/dev/null)"
+SHELLCHECK="$(command -v shellcheck 2>/dev/null)"
```
Comment (nixpanic) [inline tests/020-test-shellscripts.sh]: This change does not need to be part of this PR. I'm not sure if it is in a separate patch... Reviewing on itHub makes it difficult to make that distinction.

File: tests/300-db-import-export.sh [hunk 13]
```diff
@@ -23,7 +23,7 @@ start_server() {
 }
 
 restart_server() {
-	./heketi-server --config="./heketi.json" &>> heketi.log &
+	./heketi-server --config="./heketi.json" &> heketi.log &
```
Comment (nixpanic) [inline tests/300-db-import-export.sh]: This change does not need to be part of this PR. I'm not sure if it is in a separate patch... Reviewing on itHub makes it difficult to make that distinction.

File: apps/glusterfs/snapshot_entry.go (hunk 5 above)
Comment (MartinForReal) [reply apps/glusterfs/snapshot_entry.go]: added

File: client/api/go-client/snapshot.go (hunk 6 above)
Comment (MartinForReal) [reply client/api/go-client/snapshot.go]: added

File: client/cli/go/cmds/snapshot.go (hunk 10 above)
Comment (MartinForReal) [reply client/cli/go/cmds/snapshot.go]: yes

**Comment by MartinForReal** [comment]
Rebased on master and  deliver code change to address @nixpanic 's comments.

**Comment by MartinForReal** [comment]
added docs for snapshot command in man page

**Comment by obnoxxx** [comment]
Thanks a lot for your contribution, @MartinForReal !

@nixpanic, how does this relate to our previous started (but stalled) efforts of implementing snapshots? Is it the same direction? If so, more than happy to have it supersede that incomplete start.. :-)

**Comment by MartinForReal** [comment]
Hi @nixpanic, could you please give me some advice? Thanks!

**Comment by MartinForReal** [comment]
@nixpanic Could you please give me some advice? Thanks in advance

**Comment by amarts** [comment]
Team, can we see what more is required on this? Can we take these changes in?

**Comment by phlogistonjohn** [comment]
PR has been stale for nearly a year. Closing. Please feel free to reopen if needed.

**Pull request closed without merging**